### PR TITLE
Add `oc secrets unlink`

### PR DESCRIPTION
**Pull request by sgallagher**
New command to remove the link between a service account and a
Secret object.

This patch refactors add_secret_to_obj.go to provide several
common routines that are also used by remove_secret_from_obj.go

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

**Comment by sgallagher** [comment]
@fabianofranz @liggitt PTAL

**Comment by sgallagher** [comment]
bump

**Comment by kargakis** [comment]
@fabianofranz Instead of adding in `oc secrets` why don't we add a subcommand in `oc unset`?

**Comment by sgallagher** [comment]
@kargakis I assume that was meant to be directed at me.

I don't know if this disagrees with established patterns in Origin/Kubernetes, but having the action and its reverse be subcommands of the same command seems more discoverable to me. Also, it was one of the three proposed approaches listed in https://trello.com/c/38bAj7iu/632-1-remove-secret-from-sa (the one that I felt read most naturally).

**Comment by fabianofranz** [comment]
@kargakis I agree, and there is a [separate issue](https://github.com/openshift/origin/issues/9087) to handle that. Needs discussion usptream first as per @smarterclayton's comment.

**Comment by fabianofranz** [comment]
[test]

File: pkg/cmd/cli/secrets/remove_secret_from_object.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+const (
+	RemoveSecretRecommendedName = "remove"
+
+	// TODO: move to examples
+	removeSecretLong = `
+Remove (detach) secrets from a ServiceAccount
+If a secret is no longer valid for a pod, build or image pull, you may remove it from a service account.`
+
+	removeSecretExample = ` // Remove a secret currently associated with a service account like this:
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: We're using `#` for comments in examples.

File: pkg/cmd/cli/secrets/remove_secret_from_object.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+%[1]s serviceaccount/sa-name secret/secret-name secret/another-secret-name ...`
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: Needs "tabbing" of 2 spaces.

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 1 above)
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: Also remove the part " like this:".

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 2 above)
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: Also remove the "...", they are meant to be used only in the cmd usage. Examples should be working commands (as long as you had the resource names of course). ;)

Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: If the only target kind supported is a service account, why do you need the `serviceaccount/` part? Same for the secrets, if it's the only kind supported why `secret/`?

File: pkg/cmd/cli/secrets/remove_secret_from_object.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+type RemoveSecretOptions struct {
+	SecretOptions
+}
+
+// NewCmdRemoveSecret creates a command object for detaching one or more secret references from a service account
+func NewCmdRemoveSecret(name, fullName string, f *kcmdutil.Factory, out io.Writer) *cobra.Command {
+	o := &RemoveSecretOptions{SecretOptions{Out: out}}
+
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s serviceaccount/sa-name secret/secret-name [secret/another-secret-name] ...", name),
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: See the comment above about the need of resource names in the arguments, but I'd say the usage for this command should be something like

`oc secret remove SERVICEACCOUNT SECRET [SECRET ...] [options]`

**Comment by fabianofranz** [comment]
@smarterclayton PTAL

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: I don't think it's clear that `oc secret remove` removes the secret from a service account.  It's possible we chose `add` poorly.  `link` and `unlink` would be appropriate.  However, for consistency we can use `remove`.

**Comment by smarterclayton** [comment]
Given the lack of clarity on unset upstream this makes sense to me.

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 2 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: I have that because it's a direct mirror of the `oc secrets add` format. They require it there, so `remove` does too. Consistency is more important than brevity, IMHO.

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 3 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: @smarterclayton I could probably work up a patch to alias `add` to `link` and `remove` to unlink if you want me to. We can deprecate `add` but leave it there for compatibility.

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 2 above)
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: Yikes, you are right. `oc secrets add` is inconsistent with everything else in that sense, but `remove` should follow.

File: pkg/cmd/cli/secrets/remove_secret_from_object.go (hunk 3 above)
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/remove_secret_from_object.go]: +1, so a couple things to be fixed in terms of consistency and cleanliness: 

* `add|remove` becomes `link|unlink` (or `attach|detach`, `connect|disconnect`, `couple|decouple`)
* remove the need of `kind/` in arguments (unless for any reason we foresee supporting more than one kind for either the service account or the secret(s)).

**Comment by sgallagher** [comment]
@fabianofranz @smarterclayton PTAL

This set of patches should address all of the review comments.

**Comment by smarterclayton** [comment]
I somewhat casually suggested link as the verb, but is everyone happy with that?  Can we think of a reason not to use link?

**Comment by sgallagher** [comment]
@smarterclayton 'link' made the most sense to me out of the options suggested. I agree that "add/remove" gave the wrong impression. "attach/detach" and "connect/disconnect" seemed to me like they were transient actions, which this wouldn't be. That left "link/unlink" and "couple/decouple". "link/unlink" are easier to say and to spell, so I went with them.

**Comment by sgallagher** [comment]
[re-test]

**Comment by smarterclayton** [comment]
@ncdc how do you feel about link here?  It's a linking action for sure, and
logically if we had another link command they would feel similar (it's ok
to have two link verbs as long as they're aligned)

**Comment by ncdc** [comment]
Seems reasonable to me. Although I'd been more thinking of a generic verb such as `kubectl link <source> <target>` instead of a subcommand per source.

**Comment by smarterclayton** [comment]
The generic command wouldn't be able to handle some of the implications
this one handles (i don't think).  So we'd have to have multiple commands
anyway.

File: pkg/cmd/cli/secrets/common_secret_object.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
+package secrets
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/common_secret_object.go]: call this options.go

File: pkg/cmd/cli/secrets/common_secret_object.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+	Mapper          meta.RESTMapper
+	Typer           runtime.ObjectTyper
+	ClientMapper    resource.ClientMapper
+	ClientInterface client.Interface
+
+	Out io.Writer
+}
+
+func (o *SecretOptions) Complete(f *kcmdutil.Factory, args []string, typeFlags []string) error {
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/common_secret_object.go]: Godoc on all of these.

File: pkg/cmd/cli/secrets/link_secret_to_obj.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	SecretOptions
+
+	ForMount bool
+	ForPull  bool
+}
+
+// NewCmdLinkSecret creates a command object for linking a secret reference to a service account
+func NewCmdLinkSecret(name, fullName string, f *kcmdutil.Factory, out io.Writer) *cobra.Command {
+	o := &LinkSecretOptions{SecretOptions{Out: out}, false, false}
+	var typeFlags []string
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: For all of these

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+				found = true
+
+				// Unlink this entry from ImagePullSecrets
+				serviceaccount.ImagePullSecrets = append(serviceaccount.ImagePullSecrets[:i], serviceaccount.ImagePullSecrets[i+1:]...)
+			}
+		}
+
+		if !found {
+			// This secret didn't appear in either secret type
+			return fmt.Errorf("%s is not associated with service account %s",
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: If you are going to error on deletions where the thing doesn't exist, you need a `--force` flag so that people can script deletion.  Generally we don't error on deletions that don't exist today, so this would be inconsistent with our other commands.

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+			if secretname == serviceaccount.Secrets[i].Name {
+				found = true
+
+				// Unlink this entry from Secrets
+				serviceaccount.Secrets = append(serviceaccount.Secrets[:i], serviceaccount.Secrets[i+1:]...)
+			}
+		}
+
+		// Check the image pull secrets
+		for i := len(serviceaccount.ImagePullSecrets) - 1; i >= 0; i-- {
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: Don't do this the clever way, just create a new array and do a `for range` and append if it matches, then assign at the end if found == true.

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go (hunk 7 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: @smarterclayton I had originally discussed this with @pweil- in person. The reason I did it that way was to be protective of users making mistakes. If someone mistyped the name of a secret and received no error back from the CLI, then they're going to assume all is well. Meanwhile, the real secret is still exposed to the container.

For some linkages, I think the silent failure is fine, but when dealing with things like passwords, there's probably a really good reason why the admin is opting to unlink them and I think we probably don't want to quietly leave them linked.

If you agree, I'll work on the `--force` option.

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	if err = o.unlinkSecretsFromServiceAccount(serviceaccount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// unlinkSecretsFromServiceAccount detatches pull and mount secrets from the service account.
```
Comment (Miciah) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: Typo: "detatches".

File: pkg/cmd/cli/secrets/link_secret_to_obj.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+const (
+	// LinkSecretRecommendedName `oc secrets link`
+	LinkSecretRecommendedName = "link"
+
+	// AddSecretRecommendedName Maintained for backwards-compatibility, but deprecated
+	AddSecretRecommendedName = "add"
+
+	// TODO: move to examples
+	linkSecretLong = `
+Link secrets to a ServiceAccount
```
Comment (Miciah) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: Should it be "service account" here for consistency since you use lower-case "secrets" in the same sentence and "service account" below?

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go (hunk 7 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: OK, on reflection, I've opted to just go with the traditional approach and drop the check for unknown names.

File: pkg/cmd/cli/secrets/link_secret_to_obj.go (hunk 10 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: I've made that change here and a couple other places for consistency.

File: docs/generated/oc_by_example_content.adoc [hunk 11]
```diff
@@ -2003,20 +2003,39 @@ Change the number of pods in a deployment
...
-Add secrets to a ServiceAccount
+Link secrets to a ServiceAccount (deprecated)
 
 ====
 
 [options="nowrap"]
 ----
-  // To use your secret inside of a pod or as a push, pull, or source secret for a build, you must add a 'mount' secret to your service account like this:
-  oc secrets add serviceaccount/sa-name secrets/secret-name secrets/another-secret-name
+  # To use your secret inside of a pod or as a push, pull, or source secret for a build, you must link a 'mount' secret to your service account like this:
```
Comment (smarterclayton) [inline docs/generated/oc_by_example_content.adoc]: *technically*, you don't have to.  You only have to if the admin sets that up.

File: pkg/cmd/cli/secrets/link_secret_to_obj.go [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+	LinkSecretRecommendedName = "link"
+
+	// AddSecretRecommendedName Maintained for backwards-compatibility, but deprecated
+	AddSecretRecommendedName = "add"
+
+	// TODO: move to examples
+	linkSecretLong = `
+Link secrets to a service account
+
+After you have created a secret, you probably want to make use of that secret inside of a pod, for a build, or as an image pull secret.  In order to do that, you must link your secret to a service account.`
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: You want to hit three points here:

1. You can auto mount secrets for applications (via the mount) (uncommon)
2. You can add secrets for image pulling (common)
3. Some administrators may require secrets be linked to service accounts (v. uncommon)

So the description needs to reflect that (and the examples should also indicate a realistic example).

File: docs/generated/oc_by_example_content.adoc (hunk 11 above)
Comment (sgallagher) [inline docs/generated/oc_by_example_content.adoc]: Sorry, I don't entirely understand. How would they use this secret otherwise? Can you suggest an alternative phrasing?

File: pkg/cmd/cli/secrets/link_secret_to_obj.go (hunk 12 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: @smarterclayton I'll happily take suggestions on how better to phrase this, but this was actually existing documentation. I didn't change it.

**Comment by sgallagher** [comment]
Previous commits were missing `make update` run. Updated.

File: pkg/cmd/cli/secrets/link_secret_to_obj.go (hunk 12 above)
Comment (sgallagher) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: @smarterclayton 
The documentation changes are secondary to the functionality. I've opened https://github.com/openshift/origin/issues/9912 to track fixing that up as a follow-up.

**Comment by sgallagher** [comment]
[test]

**Comment by smarterclayton** [comment]
Documentation changes are not secondary to functionality.  You're in there
fixing up the command, but the docs need to be fixed as well.

On Mon, Jul 18, 2016 at 4:10 PM, Stephen Gallagher <notifications@github.com

**Comment by smarterclayton** [comment]
Ultimately these help snippets are the single most important bit of docs in
the project - generally everyone who improves / alters a command is
expected to revise and correct the docs to communicate the core concepts at
play to the user at the point in time the change is made.

File: pkg/cmd/cli/secrets/link_secret_to_obj.go (hunk 12 above)
Comment (liggitt) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: > You can auto mount secrets for applications (via the mount)

that's not a thing that happens today

**Comment by sgallagher** [comment]
@smarterclayton I've reworked the help snippets, PTAL.

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+const (
+	UnlinkSecretRecommendedName = "unlink"
+
+	// TODO: move to examples
+	unlinkSecretLong = `
+Unlink (detach) secrets from a service account
+If a secret is no longer valid for a pod, build or image pull, you may unlink it from a service account.`
+
+	unlinkSecretExample = ` # Unlink a secret currently associated with a service account like this:
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: No need for "like this".

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	UnlinkSecretRecommendedName = "unlink"
+
+	// TODO: move to examples
+	unlinkSecretLong = `
+Unlink (detach) secrets from a service account
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: Add a line break below this one.

**Comment by sgallagher** [comment]
@fabianofranz PTAL

**Comment by sgallagher** [comment]
@smarterclayton @fabianofranz *bump* PTAL

File: pkg/cmd/cli/secrets/unlink_secret_from_object.go (hunk 14 above)
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/unlink_secret_from_object.go]: I was expecting a bit more description.  I don't think the one line here tells me *why* I link or unlink

File: pkg/cmd/cli/secrets/link_secret_to_obj.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+Linking a secret enables a service account to automatically use that secret for some forms of authentication`
+
+	linkSecretExample = `  # Add an image pull secret to a service account to automatically use it for pulling pod images:
+  %[1]s serviceaccount-name pull-secret --for=pull
+
+  # Add an image pull secret to a service account to automatically use it for both pulling and pushing build images:
+  %[1]s builder builder-image-secret --for=pull,mount
+
+  " If the cluster's serviceAccountConfig is operating with limitSecretReferences: True, secrets must be added to the pod's service account whitelist in order to be available to the pod:
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: I believe you meant `#` instead of `"`.

File: pkg/cmd/cli/secrets/link_secret_to_obj.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+		},
+	}
+
+	cmd.Flags().StringSliceVar(&o.typeFlags, "for", []string{"mount"}, "type of secret to link: mount or pull")
+
+	return cmd
+}
+
+// NewCmdAddSecret creates a command object for linking a secret reference to a service account
```
Comment (fabianofranz) [inline pkg/cmd/cli/secrets/link_secret_to_obj.go]: Instead of having two `NewCmd*` and creating two commands to keep backwards compatibility, you should use `moved(...)` from `cli.go` like in [here](https://github.com/sgallagher/origin/blob/09101fa194f3b2109e2c632b9cc2e18ba9a5b308/pkg/cmd/cli/cli.go#L181). It takes care of adding the "DEPRECATED" message, hiding the deprecated command in help, and allows you to only have one command built, `NewCmdLinkSecret`.

**Comment by sgallagher** [comment]
@fabianofranz @smarterclayton I believe I have resolved all of the issues you identified in this latest set of patches.

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
flake, re[test]

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
flake [test]

**Comment by mfojtik** [comment]
@fabianofranz can you please do final review for this and tag this for merge (ideally today, because this one is on the 3.3 schedule).

**Comment by mfojtik** [comment]
(this LGTM, but I was hoping for at least one green test)

**Comment by sgallagher** [comment]
Actually, give me about half an hour. I realized there was one edge-case I didn't have proper handling for and I'm just fixing it up.

**Comment by sgallagher** [comment]
By which I apparently mean five minutes, because I spotted my mistake as soon as I paged out and then back in...

**Comment by deads2k** [comment]
Cute branch name.  Could use a squash.

**Comment by sgallagher** [comment]
@deads2k I was *wondering* when someone would get that reference.

**Comment by sgallagher** [comment]
@deads2k Originally, I was hoping to keep them all separate, but at this point they're so interdependent that I suppose it's fine if I squash it down to one patch. Mostly they were separate to simplify review.

**Comment by fabianofranz** [comment]
re[test] while I do a final review.

**Comment by sgallagher** [comment]
@mfojtik Tests all green!

**Comment by fabianofranz** [comment]
[merge]
